handlers/handler-router/src/main/java/org/apache/servicecomb/router/cache/RouterRuleCache.java

```diff
@@ -0,0 +1,115 @@
+import com.google.common.collect.Interners;
+import com.netflix.config.DynamicPropertyFactory;
+import com.netflix.config.DynamicStringProperty;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.servicecomb.router.model.PolicyRuleItem;
+import org.apache.servicecomb.router.model.ServiceInfoCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+import org.yaml.snakeyaml.Yaml;
+
+/**
+ * @Author GuoYl123
+ * @Date 2019/10/17
+ **/
+public class RouterRuleCache {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RouterRuleCache.class);
+
+  private static ConcurrentHashMap<String, ServiceInfoCache> serviceInfoCacheMap = new ConcurrentHashMap<>();
+
+  private static final String ROUTE_RULE = "servicecomb.routeRule.%s";
+
+  private static Interner<String> servicePool = Interners.newWeakInterner();
+
+  /**
+   * cache and register callback return false when: 1. parsing error 2. rule is null
+   *
+   * @param targetServiceName
+   * @return
+   */
+  public static boolean doInit(String targetServiceName) {
+    if (!serviceInfoCacheMap.containsKey(targetServiceName)) {
+      synchronized (servicePool.intern(targetServiceName)) {
+        if (serviceInfoCacheMap.containsKey(targetServiceName)) {
+          return true;
+        }
+        //Yaml not thread-safe
+        Yaml yaml = new Yaml();
+        DynamicStringProperty ruleStr = DynamicPropertyFactory.getInstance().getStringProperty(
+            String.format(ROUTE_RULE, targetServiceName), null, () -> {
+              Yaml temYaml = new Yaml();
+              refresh(targetServiceName);
+              DynamicStringProperty tepRuleStr = DynamicPropertyFactory.getInstance()
+                  .getStringProperty(String.format(ROUTE_RULE, targetServiceName), null);
+              addAllRule(targetServiceName, temYaml, tepRuleStr);
+            });
+        return addAllRule(targetServiceName, yaml, ruleStr);
+      }
+    }
+    return true;
+  }
+
+  private static boolean addAllRule(String targetServiceName, Yaml yaml,
+      DynamicStringProperty ruleStr) {
+    if (ruleStr.get() == null) {
+      return false;
+    }
+    List<PolicyRuleItem> policyRuleItemList = Arrays
+        .asList(yaml.loadAs(ruleStr.get(), PolicyRuleItem[].class));
+    if (CollectionUtils.isEmpty(policyRuleItemList)) {
+      return false;
+    }
+    try {
+      if (serviceInfoCacheMap.get(targetServiceName) == null) {
+        serviceInfoCacheMap.put(targetServiceName, new ServiceInfoCache());
+      }
+      serviceInfoCacheMap.get(targetServiceName).setAllrule(policyRuleItemList);
+      // init tagitem
+      serviceInfoCacheMap.get(targetServiceName).getAllrule().forEach(a ->
+          a.getRoute().forEach(b -> b.initTagItem())
+      );
+      // sort by precedence
+      serviceInfoCacheMap.get(targetServiceName).sortRule();
```

Comment:
How about initialize ServiceInfoCache first and then put it to serviceInfoCacheMap? or the sysnchronized will get a empty list 
```
   if (!serviceInfoCacheMap.containsKey(targetServiceName)) {
```
Although this is not a very bit problem .

Reply:
has been refactory